eth/consensus/clique/snapshot_manager.py

```diff
@@ -0,0 +1,311 @@
+    is_checkpoint,
+)
+
+
+def make_snapshot_lookup_key(block_hash: Hash32) -> bytes:
+    return b'block-hash-to-snapshot:%s' % block_hash
+
+
+class SnapshotManager:
+    """
+    The ``SnapshotManager`` is responsible for managing the snapshots that hold the current
+    state of the consensus engine. It creates new snapshots by applying headers on top of
+    existing snapshots. It comes with APIs to create, persist and retrieve snapshots.
+    """
+
+    logger = get_extended_debug_logger('eth.consensus.clique.snapshot_manager.SnapshotManager')
+
+    def __init__(self,
+                 chain_db: ChainDatabaseAPI,
+                 header_cache: HeaderCache,
+                 epoch_length: int) -> None:
+        self._snapshots = lru.LRU(IN_MEMORY_SNAPSHOTS)
+        self._header_cache = header_cache
+        self._chain_db = chain_db
+        self._epoch_length = epoch_length
+
+    def _lookup_header(self, block_hash: Hash32) -> BlockHeader:
+
+        if block_hash in self._header_cache:
+            return self._header_cache[block_hash]
+
+        try:
+            return self._chain_db.get_block_header_by_hash(block_hash)
+        except HeaderNotFound:
+            raise ValidationError("Unknown ancestor %s", block_hash)
+
+    def _create_snapshot_from_checkpoint_header(self, header: BlockHeader) -> Snapshot:
+        signers = get_signers_at_checkpoint(header)
+        self.logger.debug2("Created snapshot from checkpoint at %s", header)
+
+        snapshot = MutableSnapshot(
+            signers=list(signers),
+            block_hash=header.hash,
+            votes=[],
+            tallies={}
+        )
+        return self.add_snapshot(snapshot)
+
+    def apply(self, current_snapshot: Snapshot, header: BlockHeader) -> Snapshot:
+        """
+        Apply the given header on top of the current snapshot to create a new snapshot.
+        """
+
+        if is_checkpoint(header.block_number, self._epoch_length):
+            return self._create_snapshot_from_checkpoint_header(header)
+
+        snapshot = current_snapshot.get_mutable_clone(header.hash)
+
+        if header.nonce in {NONCE_AUTH, NONCE_DROP}:
+            signer = get_block_signer(header)
+            # Clear any votes from the signer regarding the subject that is voted on
+            for vote in snapshot.votes:
+                if vote.signer == signer and vote.subject == header.coinbase:
+                    self.uncast(snapshot, vote.subject, vote.action)
+                    snapshot.votes.remove(vote)
+                    break
+
+            action = VoteAction.NOMINATE if header.nonce == NONCE_AUTH else VoteAction.KICK
+
+            if self.cast(snapshot, header.coinbase, action):
+                vote = Vote(
+                    signer=signer,
+                    block_number=header.block_number,
+                    subject=header.coinbase,
+                    action=action
+                )
+                snapshot.votes.append(vote)
```

Comment:
You might consider using a datatype from [`pyrsistent`](https://pypi.org/project/pyrsistent/) instead and using their Evolver API to do this mutation.